src/components/views/rooms/UserIdentityWarning.tsx

```diff
@@ -0,0 +1,321 @@
+        // anything.  If encryption gets enabled later, we will retry, via
+        // onRoomStateEvent.
+        if (!(await crypto.isEncryptionEnabledInRoom(room.roomId))) {
+            return;
+        }
+        initialisedRef.current = InitialisationStatus.Initialising;
+
+        const members = await room.getEncryptionTargetMembers();
+
+        for (const member of members) {
+            await addMemberIfNeedsApproval(member.userId, member);
+        }
+
+        setCurrentPrompt(selectCurrentPrompt());
+        initialisedRef.current = InitialisationStatus.Completed;
+    }, [crypto, room, addMemberIfNeedsApproval, selectCurrentPrompt]);
+
+    loadMembers().catch((e) => {
+        logger.error("Error initialising UserIdentityWarning:", e);
+    });
+
+    // When a user's verification status changes, we check if they need to be
+    // added/removed from the set of members needing approval.
+    const onUserVerificationStatusChanged = useCallback(
+        (userId: string, verificationStatus: UserVerificationStatus): void => {
+            const gotVerificationStatusUpdate = gotVerificationStatusUpdateRef.current;
+
+            // If we haven't started initialising, that means that we're in a
+            // room where we don't need to display any warnings.
+            if (initialisedRef.current === InitialisationStatus.Uninitialised) {
+                return;
+            }
+
+            if (gotVerificationStatusUpdate.has(userId)) {
+                gotVerificationStatusUpdate.set(userId, true);
+            }
+
+            if (verificationStatus.needsUserApproval) {
+                addMemberNeedingApproval(userId);
+            } else {
+                removeMemberNeedingApproval(userId);
+            }
+        },
+        [addMemberNeedingApproval, removeMemberNeedingApproval],
+    );
+    useTypedEventEmitter(cli, CryptoEvent.UserTrustStatusChanged, onUserVerificationStatusChanged);
+
+    // We watch for encryption events (since we only display warnings in
+    // encrypted rooms), and for membership changes (since we only display
+    // warnings for users in the room).
+    const onRoomStateEvent = useCallback(
+        async (event: MatrixEvent): Promise<void> => {
+            if (!crypto || event.getRoomId() !== room.roomId) {
+                return;
+            }
+
+            const eventType = event.getType();
+            if (eventType === EventType.RoomEncryption && event.getStateKey() === "") {
+                // Room is now encrypted, so we can initialise the component.
+                return loadMembers().catch((e) => {
+                    logger.error("Error initialising UserIdentityWarning:", e);
+                });
+            } else if (eventType !== EventType.RoomMember) {
+                return;
+            }
+
+            if (initialisedRef.current === InitialisationStatus.Uninitialised) {
+                return;
+            }
```

Comment:
this means that any membership changes that arrive while we're figuring out the approval status of the initial member list will be ignored? That seems wrong?

Reply:
If we're figuring out the approval status of the initial member list, then `initialisedRef.current` will be `InitialsationStatus.Initialising`, so we won't return here.

Reply:
ah, got it.